File: leetcode/algorithms/maximum_gap/maximum_gap.cc
class Solution {
public:
    int maximumGap(vector<int>& nums) {
        int n = nums.size();

        if (n < 2) {
            return 0;
        }

        int min_value = *min_element(nums.begin(), nums.end());
        int max_value = *max_element(nums.begin(), nums.end());

        int gap = (max_value - min_value - 1) / (n - 1) + 1;

        if (gap == 0) {
            return 0;
        }

        vector<pair<int, int>> buckets(n, make_pair(INT_MAX, INT_MIN));

        for (int &value: nums) {
            int bucket = (value - min_value) / gap;

            int bucket_min = min(buckets[bucket].first, value);
            int bucket_max = max(buckets[bucket].second, value);

            buckets[bucket] = make_pair(bucket_min, bucket_max);
        }

        int max_gap = INT_MIN;
        int prev = min_value;

        for (pair<int, int> &bucket: buckets) {
            if (bucket.first == INT_MAX && bucket.second == INT_MIN) {
                continue;
            }

            max_gap = max(max_gap, bucket.first - prev);
            prev = bucket.second;
        }

        return max_gap;
    }
};

File: leetcode/algorithms/maximum_gap/maximum_gap.py
class Solution:
    def maximumGap(self, values):
        n = len(values)

        if n < 2:
            return 0

        min_value = min(values)
        max_value = max(values)

        gap = float(max_value - min_value) / (n - 1)

        if gap == 0:
            return 0

        buckets = [[float('inf'), 0] for __ in xrange(n)]

        for value in values:
            bucket = int((value - min_value) / gap)

            bucket_min = min(buckets[bucket][0], value)
            bucket_max = max(buckets[bucket][1], value)

            buckets[bucket][0] = bucket_min
            buckets[bucket][1] = bucket_max

        max_gap = 0
        prev = min_value

        for bucket_min, bucket_max in buckets:
            if bucket_min == float('inf') and bucket_max == 0:
                continue

            max_gap = max(max_gap, bucket_min - prev)
            prev = bucket_max

        return max_gap
